src/globus_cli/services/transfer/recursive_ls.py

```diff
@@ -21,63 +60,57 @@ class RecursiveLsResponse:
-          ``endpoint_id``
-            The endpoint that will be recursively ls'ed.
-          ``max_depth``
-            The maximum depth the recursive ls will go into the file system.
-          ``filter_after_first``
-            If True, any filter in ``ls_params`` will be applied to all calls
-            If False, any filter in ``ls_params`` will only be applied to the
-            first ls.
-          ``ls_params``
-            Query params sent to operation_ls, see operation_ls for more
-            details.
-        """
-        logger.info(
-            "Creating RecursiveLsResponse on path {} of endpoint {}".format(
-                ls_params.get("path"), endpoint_id
-            )
+    def __init__(
+        self,
+        client: TransferClient,
+        endpoint_id: str,
+        ls_params: Dict[str, Any],
+        *,
+        max_depth: int = 3,
+        filter_after_first: bool = True,
+    ) -> None:
+        self._client = client
+        self._endpoint_id = endpoint_id
+        self._ls_params = ls_params
+        self._max_depth = max_depth
+        self._filter_after_first = filter_after_first
+
+        start_path = cast(Optional[str], ls_params.get("path"))
+        log.info(
+            "Creating RecursiveLsResponse on path '%s' of endpoint '%s'",
+            start_path,
+            endpoint_id,
         )
 
-        self.client = client
-        self.endpoint_id = endpoint_id
-        self.ls_params = ls_params
-        self.max_depth = max_depth
-        self.filter_after_first = filter_after_first
-        self.filtering = True
-        self.ls_count = 0
-
-        # queue of (absolute_path, relative_path, depth) tuples.
-        self.queue = deque()
-        # initialized with the start path (if any) and a depth of 0
-        self.queue.append((self.ls_params.get("path"), "", 0))
-
         # call the iterable_func method to convert it to a generator expression
-        self.generator = self.iterable_func()
+        self._generator = self._iterable_func(start_path)
 
         # grab the first element out of the internal iteration function
         # because this could raise a StopIteration exception, we need to be
         # careful and make sure that such a condition is respected (and
         # replicated as an iterable of length 0)
         try:
-            self.first_elem = next(self.generator)
+            self._first_elem: Optional[ITEM_T] = next(self._generator)
         except StopIteration:
-            # express this internally as "generator is null" -- just need some
+            # express this internally as "first_elem is null" -- just need some
             # way of making sure that it's clear
-            self.generator = None
+            self._first_elem = None
 
-    def __iter__(self):
-        yield self.first_elem
-        yield from self.generator
+    def __iter__(self) -> Iterator[ITEM_T]:
+        if self._first_elem is not None:
+            yield self._first_elem
```

Comment:
I'm wondering here what this first elem thing is accomplishing—why not just `yield from self._generator`?

Reply:
The first element is grabbed from the generator eagerly so that an error will be triggered immediately if the first network call fails.
Otherwise, it's possible to have semantics like
```python
res = recursive_ls(...)  # passes
print(res)  # fails
```

It's still possible to start failing part-way through, but this ensures that once the initial `recursive_ls()` function returns, at least one call worked.